IntelliJ IDEA: Change Setup Instructions #7322

LiHaoTan opened the pull request:
Closes #7322

**Checklist**

- [x] Automated Setup Docs
- [x] Manual Setup Docs
- [x] Automated Setup Behavior Docs
- [x] Depends on #7361 to be merged
- [x] Revert #6918

**Outline of Solution**
1. Use single module instead of multiple modules, so only one build directory is used.
2. Add automated setup for IntelliJ.
3. Change instructions for manual setup.
3. Add IntelliJ specific Gradle task for JDO enhancement.

**Consequent issues closed/fixed**
Fixes #6972 by changing the build output path to differ from Gradle. In addition, instructions are added for the user to put their datastore path in another folder. More info [here](https://github.com/TEAMMATES/teammates/pull/7354/commits/c871380887d591bf1269c38252213a9e48df41a8)

#6919 was resolved previously but because I could not reach the original contributor it becomes a mystery. I do believe it to be a user error though because the instructions are easy to miss. I have not thought of a good way to address that in the manual setup but the automated setup would not have such an issue.

Thus [docs/troubleshooting-guide.md](https://github.com/TEAMMATES/teammates/pull/6918/files#diff-7bb9f7f577d1c74e5d6259ba1d063058) is reverted even though I was not able to confirm the cause because the solution is not right. There is nothing wrong with the `Output directory` path even if the cause is (which is unlikely unless there is an IntelliJ bug). The path may also not persist because the Gradle build is the "source of truth" and so the manual setting can get reset anytime without warning. More details in 1e65b66. We can also re-troubleshoot on a case by case basis if such an issue were to occur again so we actually find the cause.

#6860 was closed by the user when the issue was not resolved. However, it was fixed by #6980. The new instructions here takes a different approach in the automated setup but the same approach as #6980 in the manual setup.

Fixes #6702
* Configure → Project Defaults → Project Structure
* Click Import existing project and select the local repository folder
> Both of the instructions forces the user to be on the welcome screen and force defaults on the user they may not want. The new instructions do not depend on the user being on the welcome screen.

This menu path is not present in the main UI
![image](https://cloud.githubusercontent.com/assets/13617389/26459123/08532534-41a8-11e7-90f8-409a1246b1d0.png)
> assemble is no longer required in the new instructions

* Go to Run → Play
Use Run... instead of Play?
> Already closed by #6764

* Click + → .... Select the App Engine SDK you downloaded in Step 3 of the Setting up a development
environment guide.
> Already closed by #6809

Fixes #7158 by having only one build folder. Previously there was `out` and `build` which can cause the error. Issue further described in 	1e65b66.

Fixes #6971
There are two issues at play here.
1. Build directory is wrong. `IntelliJ` and `Gradle` (or rather `appengine` because Gradle does not explode wars by default) actually uses different exploded war directories. Further details in 1e65b66.
2. In some sense, because a `Gradle` task is used to enhance the classes instead of using IntelliJ Google App Engine integration, the directory for the exploded war need to match the one in the `appengine` plugin. However, it is described in 1e65b66 the problems with that.

Instead of using a `Gradle` task to enhance, it is better to use the Google App Engine integration in IntelliJ to perform the enhancement but it is not working properly, thus a new `Gradle` task specifically for enhancing in IntelliJ is created. I was not able to find out why the enhancement in Google App Engine integration was only working partially but it is not worth to investigate further due to the upcoming `Objectify` migration, so it is to be noted this is a temporary approach.

**Dependent issue closed**
Closes #7288 as the automated setup sets up Stylelint automatically. Thus manual instructions for setting up Styleint is also added. While issue #7288 could be separated and addressed with a different PR there is coupling between the setup instructions and the issue so it will require additional unneeded effort to separate them.

**Caveats**
* Single module instead of multiple modules for the project is used.

   While multiple modules has benefits in separating the dependencies, I was not able to come up with a good solution to support it.

    According to Jetbrains there are also additional benefits in that multiple modules per source set solves some [problems](https://blog.jetbrains.com/idea/2016/03/intellij-idea-2016-1-is-here/). However, multiple modules per source set also caused a number of [problems](https://stackoverflow.com/questions/36330542/control-automatic-creation-of-idea-modules-from-gradle-sourcesets-in-2016-1).

   Two solutions were considered for using multiple modules:
   1. Generating project files through the `idea`plugin in Gradle. However, the plugin is still lacking and [does not yet support](https://goo.gl/tx5lJ2) generating the better directory-based project.
  2. Editing the project files after they are created by the IntelliJ. However, I was not able to come up with a clean solution for this. See 1e65b66 and b113b12.
* IntelliJ JDO enhancement uses an unclean approach. (Will be resolved after Objectify migration since enhancement is no longer needed)
* ES6 Javascript files are not transpiled to ES5 in IntelliJ automatically.
* Include in project build is easy to miss if they did not read [docs/development.md](https://github.com/TEAMMATES/teammates/blob/master/docs/development.md#managing-the-dev-server).

LiHaoTan commented:
Ready for review

damithc commented:
Tried the instructions. Generally went well. Some nits:

1. add `as follows:` here?. At first it wasn't clear to me that the instructions are given below.
![image](https://cloud.githubusercontent.com/assets/1673303/26719014/7e6494d4-47b5-11e7-845d-c587325ebbb5.png)

2. `Gradle VM Options` should be in code format?
![image](https://cloud.githubusercontent.com/assets/1673303/26719164/0adbb2b2-47b6-11e7-9f69-0c94b4bb0020.png)

3. Got this warning when opening the newly-created project in Intellij. Not sure it's due to my previous setups. Anyway, the right response seems to be 'remove'
![image](https://cloud.githubusercontent.com/assets/1673303/26719436/1b83fede-47b7-11e7-8680-2dddc1dc7400.png)

4. Step 2 _must_ be done before step 3? Also, `setupIntellijSettings` --> `setupIntellijProjectSettings`?
![image](https://cloud.githubusercontent.com/assets/1673303/26719557/98242860-47b7-11e7-928b-1a18aad88ba6.png)

LiHaoTan commented:
1. Added `as follows:` to prerequisites.
![image](https://cloud.githubusercontent.com/assets/13617389/26731173/c2f7a764-47e5-11e7-88e3-5b6e0d596607.png)

2. Added the code formatting.
![image](https://cloud.githubusercontent.com/assets/13617389/26731209/dd6e170e-47e5-11e7-82ff-207a2888b689.png)

3. I think it should be due to the previous setups, I think IntelliJ ended up duplicating the modules and they detected one is rendered obsolete.
![image](https://cloud.githubusercontent.com/assets/13617389/26731222/e6d68902-47e5-11e7-9469-c9bcf7ecf94e.png)

4. It should be in that order because I cannot configure some settings as they are user specific until they open the project, but since you already had a project configured previously, following the steps in the order of `1, 3, 2` would work as well.
![image](https://cloud.githubusercontent.com/assets/13617389/26731244/fe8af4c0-47e5-11e7-89b5-de6b4d633198.png)

Because there is a previous setup already done, I am not sure if the instructions have worked correctly or incorrectly. Do you mind if you do a fresh project setup(just delete the .`idea` folder) ?

damithc commented:
I did the setup twice, second time with a fresh clone. Strangely, the warning (item 3) popped up both times. Will be doing another set up in another computer soon. That computer doesn't have any previous setups. Will let you know if something noteworthy happens during that set up.

damithc commented:
I tried to do a fresh setup in a new computer. The warning in item 3 still showed up. May be because my root directory is not `teammates`? I did not observe any other problems.

LiHaoTan commented:
@damithc ah I did not consider a different project root. The current setup will cause some minor quirks if you used a different project root.

I added some settings that will now support a different project root.

If you refresh your project, the module with the name of `$ProjectName$` should now be asked to be removed, and you would have to re-configure the `teammates` module, you would probably receive some pop-ups like in the manual setup.

Alternatively, you can do a fresh setup and should no longer get any pop-ups.

whipermr5 commented:
@LiHaoTan Sorry for the delay! Reviewing now :)

whipermr5 commented on docs/ide-usage.md:
```diff
@@ -86,6 +86,12 @@ The following plugins are needed:
+**NOTE**
+> If you are migrating from Eclipse, you should delete `.project`, `.classpath`, `.launches`,
+> `src/main/webapp/WEB-INF/classes`, `src/main/webapp/WEB-INF/lib` and `src/main/webapp/WEB-INF/appengine-generated`.\
+> You can also copy `local_db.bin` from `src/main/webapp/WEB-INF/appengine-generated` and copy it to the appropriate
```
~~and copy it~~

damithc replied:
On a related note, I wish intellij/gradle can preserve the local db through builds/cleans. Test data previous entered should not be wiped out during builds. But probably no need to do in this PR.

whipermr5 replied:
@LiHaoTan has included instructions how to fix this here: https://github.com/TEAMMATES/teammates/pull/7354/files#diff-1bf07101845de6da98f1c948b82c5ce0R84

whipermr5 commented on docs/ide-usage.md:
```diff
@@ -86,27 +86,118 @@ The following plugins are needed:
+> If you are migrating from Eclipse, you should delete `.project`, `.classpath`, `.launches`,
+> `src/main/webapp/WEB-INF/classes`, `src/main/webapp/WEB-INF/lib` and `src/main/webapp/WEB-INF/appengine-generated`.\
+> You can also copy `local_db.bin` from `src/main/webapp/WEB-INF/appengine-generated` and copy it to the appropriate
+> location your dev server in IntelliJ reads the datastore from after your IntelliJ setup.
```
Can this location be made explicit?

whipermr5 commented on docs/ide-usage.md:
```diff
@@ -86,27 +86,118 @@ The following plugins are needed:
+      `Frameworks detected: Google App Engine, Web, JPA frameworks are detected in the project`.\
+      **OR**\
+      `Frameworks detected: OSGi, Google App Engine, Web, JPA frameworks are detected in the project`.\
+      Click `Configure`, ensure that there is only `Google App Engine` and `JPA` frameworks being shown, otherwise make
```
Click `Configure`~~,~~ **and** ensure that ~~there is~~ only `Google App Engine` and `JPA` frameworks ~~being~~ **are** shown,

whipermr5 commented on docs/staticAnalysis.md:
```diff
@@ -209,17 +282,37 @@ The link to the report will be displayed in each PR, or by clicking the badge on
+> There are some packages and classes that are excluded when using Jacoco on the comamnd line which are not excluded
+when you run them in IntelliJ, thus the coverage statistics you see in IntelliJ may not match what you see on `Codecov`. 
+
+Alternatively, you can right click a class with TestNG test(s), and click `Run '$FileClass$' with Coverage`, this will use
```
click a class with TestNG test(s)~~,~~ and click

whipermr5 commented on docs/staticAnalysis.md:
```diff
@@ -209,17 +282,37 @@ The link to the report will be displayed in each PR, or by clicking the badge on
+### CLI
 
-Alternatively, use Gradle to run the tests, and obtain the coverage data with `jacocoReport` task, i.e:
+You can use use Gradle to run tests, and obtain the coverage data with `jacocoReport` task, i.e:
```
run tests~~,~~ and

whipermr5 commented on docs/staticAnalysis.md:
```diff
@@ -199,6 +234,44 @@ To run all static analysis tasks in one sitting, run the following two commands:
+> `Analyze → Inspect Code... → Whole project`.
+
+`PMD` is provided as a plugin and do not run automatically. You can run it by selecting a number of files, right clicking,
+select `Run PMD` and then the option you want.
```
select**ing*** `Run PMD` and then **choosing** the option you want.
